File: service-chart/src/domain/input.weekly.ts
import { TimeSeriesType } from '../model';
import { LoadChartInputBase } from './input.base';

export class LoadChartInputWeekly implements LoadChartInputBase {
  timeSeriesType: Readonly<TimeSeriesType> = 'TIME_SERIES_WEEKLY';

  timeSeriesKey: Readonly<string> = 'Weekly Time Series';

  symbol: Readonly<string>;

  constructor(symbol: string) {
    this.symbol = symbol;
  }

  toDescriptor(): Readonly<string> {
    return `${this.symbol}.${this.timeSeriesType}`;
  }

  toReferenceUrl(
    prefix: Readonly<String>,
    apiKey: Readonly<string>,
  ): Readonly<string> {
    return (
      `${prefix}`
      + `?function=${this.timeSeriesType}`
      + `&symbol=${this.symbol}`
      + `&apikey=${apiKey}`
    );
  }

  toTimeSeriesKey(): Readonly<string> {
    return this.timeSeriesKey;
  }
}

File: stock-tracker/src/chart_loader/input_weekly.rs
use crate::chart_loader::{Creator, LoadChartInputTrait};

pub struct InputWeekly {
  timeseries_type: String,
  timeseries_key: String,
  symbol: String,
}

impl Creator for InputWeekly {
  fn new(symbol: String) -> Self {
    Self {
      timeseries_type: String::from("TIME_SERIES_WEEKLY"),
      timeseries_key: String::from("Weekly Time Series"),
      symbol,
    }
  }
}

impl LoadChartInputTrait for InputWeekly {
  fn to_descriptor(&self) -> String {
    format!("{}.{}", self.symbol, self.timeseries_type)
  }

  fn to_query_string(&self, prefix: String, apikey: String) -> String {
    format!(
      "{}?function={}&symbol={}&apikey={}",
      prefix, self.timeseries_type, self.symbol, apikey
    )
  }

  fn to_timeseries_key(&self) -> String {
    self.timeseries_key.clone()
  }
}
